From a127c792c974706d40674adbd795735138f59c36 Mon Sep 17 00:00:00 2001
From: Raphael Mounier <mounierr07@gmail.com>
Date: Sun, 16 Nov 2025 12:35:07 +0100
Subject: [PATCH 3/3] Don't crash when PowerHal have an old version

Fix this crash in zygote

Caused by: java.lang.NullPointerException: Attempt to read from field 'android.hardware.power.SupportInfo$HeadroomSupportInfo android.hardware.power.SupportInfo.headroom' on a null object reference in method 'void com.android.server.power.hint.HintManagerService.<init>(android.content.Context, com.android.server.power.hint.HintManagerService$Injector)'

Change-Id: Ic3a546fa2eb4251ac305f2b13a4b5ed50df58ea6
---
 .../server/power/hint/HintManagerService.java | 93 ++++++++++++++-----
 1 file changed, 71 insertions(+), 22 deletions(-)

diff --git a/services/core/java/com/android/server/power/hint/HintManagerService.java b/services/core/java/com/android/server/power/hint/HintManagerService.java
index 987a84994451..e89c7567ee77 100644
--- a/services/core/java/com/android/server/power/hint/HintManagerService.java
+++ b/services/core/java/com/android/server/power/hint/HintManagerService.java
@@ -295,23 +295,72 @@ public final class HintManagerService extends SystemService {
         mPowerHal = injector.createIPower();
         mPowerHalVersion = 0;
         mUsesFmq = false;
+
         if (mPowerHal != null) {
+            Slog.d(TAG, "PowerHal is not null");
             mSupportInfo = getSupportInfo();
         }
-        mDefaultCpuHeadroomCalculationWindowMillis =
-                new CpuHeadroomParamsInternal().calculationWindowMillis;
-        mDefaultGpuHeadroomCalculationWindowMillis =
-                new GpuHeadroomParamsInternal().calculationWindowMillis;
-        if (mSupportInfo.headroom.isCpuSupported) {
-            mCpuHeadroomCache = new HeadroomCache<>(2, mSupportInfo.headroom.cpuMinIntervalMillis);
-        } else {
-            mCpuHeadroomCache = null;
-        }
-        if (mSupportInfo.headroom.isGpuSupported) {
-            mGpuHeadroomCache = new HeadroomCache<>(2, mSupportInfo.headroom.gpuMinIntervalMillis);
-        } else {
-            mGpuHeadroomCache = null;
+        else {
+            Slog.w(TAG, "PowerHal is null - perhaps yuo have an old powerhal HIDL");
+            mSupportInfo = getDummySupportInfo();
+        }
+
+	mDefaultCpuHeadroomCalculationWindowMillis =
+	        new CpuHeadroomParamsInternal().calculationWindowMillis;
+	mDefaultGpuHeadroomCalculationWindowMillis =
+	        new GpuHeadroomParamsInternal().calculationWindowMillis;
+
+	if (mSupportInfo.headroom.isCpuSupported) {
+	    mCpuHeadroomCache = new HeadroomCache<>(2, mSupportInfo.headroom.cpuMinIntervalMillis);
+	} else {
+	    mCpuHeadroomCache = null;
+	}
+	if (mSupportInfo.headroom.isGpuSupported) {
+	    mGpuHeadroomCache = new HeadroomCache<>(2, mSupportInfo.headroom.gpuMinIntervalMillis);
+	} else {
+	    mGpuHeadroomCache = null;
+	}
+    }
+
+    private SupportInfo getDummySupportInfo() {
+        SupportInfo supportInfo = new SupportInfo();
+        supportInfo.usesSessions = isHintSessionSupported();
+        // Global boosts & modes aren't currently relevant for HMS clients
+        supportInfo.boosts = 0;
+        supportInfo.modes = 0;
+        supportInfo.sessionHints = 0;
+        supportInfo.sessionModes = 0;
+        supportInfo.sessionTags = 0;
+
+        supportInfo.headroom = new SupportInfo.HeadroomSupportInfo();
+        supportInfo.headroom.isCpuSupported = false;
+        supportInfo.headroom.isGpuSupported = false;
+
+        supportInfo.compositionData = new SupportInfo.CompositionDataSupportInfo();
+        if (isHintSessionSupported()) {
+            if (mPowerHalVersion == 4) {
+                // Assume we support the V4 hints & modes unless specified
+                // otherwise; this is to avoid breaking backwards compat
+                // since we historically just assumed they were.
+                supportInfo.sessionHints = 31; // first 5 bits are ones
+            }
+            if (mPowerHalVersion == 5) {
+                // Assume we support the V5 hints & modes unless specified
+                // otherwise; this is to avoid breaking backwards compat
+                // since we historically just assumed they were.
+
+                // Hal V5 has 8 modes, all of which it assumes are supported,
+                // so we represent that by having the first 8 bits set
+                supportInfo.sessionHints = 255; // first 8 bits are ones
+                // Hal V5 has 1 mode which it assumes is supported, so we
+                // represent that by having the first bit set
+                supportInfo.sessionModes = 1;
+                // Hal V5 has 5 tags, all of which it assumes are supported,
+                // so we represent that by having the first 5 bits set
+                supportInfo.sessionTags = 31;
+            }
         }
+        return supportInfo;
     }
 
     SupportInfo getSupportInfo() {
@@ -324,11 +373,9 @@ public final class HintManagerService extends SystemService {
             throw new IllegalStateException("Could not contact PowerHAL!", e);
         }
 
-        SupportInfo supportInfo = new SupportInfo();
-        supportInfo.headroom = new SupportInfo.HeadroomSupportInfo();
-        supportInfo.headroom.isCpuSupported = false;
-        supportInfo.headroom.isGpuSupported = false;
-        return supportInfo;
+        Slog.w(TAG, "PowerHal version < 6 - return DummySupportInfo");
+
+        return getDummySupportInfo();
     }
 
     private ServiceThread createCleanUpThread() {
@@ -1448,8 +1495,10 @@ public final class HintManagerService extends SystemService {
 
         @Override
         public CpuHeadroomResult getCpuHeadroom(@NonNull CpuHeadroomParamsInternal params) {
-            if (!mSupportInfo.headroom.isCpuSupported) {
-                throw new UnsupportedOperationException();
+	    if (mSupportInfo != null) {
+		    if (!mSupportInfo.headroom.isCpuSupported) {
+		        throw new UnsupportedOperationException();
+		    }
             }
             final CpuHeadroomParams halParams = new CpuHeadroomParams();
             halParams.tids = new int[]{Binder.getCallingPid()};
@@ -1584,8 +1633,8 @@ public final class HintManagerService extends SystemService {
                     }
                 }
             }
-            pw.println("CPU Headroom Interval: " + mSupportInfo.headroom.cpuMinIntervalMillis);
-            pw.println("GPU Headroom Interval: " + mSupportInfo.headroom.gpuMinIntervalMillis);
+            //pw.println("CPU Headroom Interval: " + mSupportInfo.headroom.cpuMinIntervalMillis);
+            //pw.println("GPU Headroom Interval: " + mSupportInfo.headroom.gpuMinIntervalMillis);
             try {
                 CpuHeadroomParamsInternal params = new CpuHeadroomParamsInternal();
                 params.usesDeviceHeadroom = true;
-- 
2.25.1

